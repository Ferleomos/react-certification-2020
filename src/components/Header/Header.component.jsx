import styled from 'styled-components';

const Header = styled.header`
    background-color: #60a4e4;
    display: flex;
    flex-direction: row;
    border: none;
    z-index: 1100;
    width: 100%;
    min-height: 64px;
    align-items: center;
    position: sticky;
    top: 0;
`;

export default Header;