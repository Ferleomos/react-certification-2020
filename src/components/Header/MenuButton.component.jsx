import React from 'react';
import styled from 'styled-components';

const Icon = styled.svg`
    flex: none;
    width: 25px;
    height: 25px;
`;

const Button = styled.button`
    background-color: transparent;
    color: transparent;
    border: none;
    display: inline-flex;
    position: relative;
    align-items: center;
    margin: 0px 25px 0px 15px;

    @media(max-width: 640px) {
        margin: 0px 0px 0px 10px;
    }
`;

function MenuIcon(){
    return(
        <Button data-testid="btnMenuIcon">
            <Icon viewBox="0 0 24 24">
                <path d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"></path>
            </Icon>
        </Button>
    );
}

export default MenuIcon;