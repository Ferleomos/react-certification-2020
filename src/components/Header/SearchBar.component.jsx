import React from 'react';
import styled from 'styled-components';

const SearchContainer = styled.div`
    border: none;
    display: inline-flex;
    width: 100%;
    margin: auto;
    margin-left: 25px;

    @media(max-width: 640px) {
        margin: 0px 10px 0px 5px;
    }

    @media(max-width: 400px) {
        justify-content: center;
    }
`;

const SearchInput = styled.input`
    width: 100%;
    max-width: 260px;
    min-height: 35px;
    max-height: 35px;
    border-radius: 4px;
    background-color: rgba(255, 255, 255, 0.60);
    border: none;
`;

function SearchBar(){
    return(
        <SearchContainer>
            <SearchInput placeholder='Search...'></SearchInput>
        </SearchContainer>
    );
};

export default SearchBar;