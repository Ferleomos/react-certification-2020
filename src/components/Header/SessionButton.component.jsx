import React from 'react';
import styled from 'styled-components';

const Icon = styled.svg`
    fill: ${(props) => props.theme.hForeground};
`;

const UserButton = styled.button`
    border: none;
    border-radius: 50%;
    height: 40px;
    min-width: 40px;
    background-color: #bdbdbd;
    color: transparent;
    display: inline-flex;
    margin: 0px 20px;
    justify-content: center;
    vertical-align: center;
    align-items: center;
    align-content: center;
    outline: none;
    
    @media(max-width: 640px) {
        display: none;
    }
`;

function SessionButton(){
    return(
        <UserButton>
            <Icon viewBox="0 0 24 24">
                <path d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z" />
            </Icon>
        </UserButton>
    );
};

export default SessionButton;