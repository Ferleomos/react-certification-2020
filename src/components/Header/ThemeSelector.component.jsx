import React from 'react';
import styled from 'styled-components';

const ThemeContainer = styled.div`
    border: none;
    display: inline-flex;
    justify-content: flex-end;
    vertical-align: center;
    width: 100%;
    margin: auto;
    align-items: center;
    padding: 15px;

    @media(max-width: 640px) {
        display: none;
    }
`;

const ThemeLabel = styled.label`
    color: Black;
    font-weight: bold;
    font-family: "Roboto", "Helvetica", "Arial", sans-serif;
`;

const ThemeButton = styled.input.attrs({type: 'checkbox'})`
   margin: 0px 15px;
`;

function ThemeSelector(){
    return(
        <ThemeContainer>
            <ThemeButton>

            </ThemeButton>
            <ThemeLabel aria-label="Dark Mode">Dark Mode</ThemeLabel>
        </ThemeContainer>
    );
};

export default ThemeSelector;