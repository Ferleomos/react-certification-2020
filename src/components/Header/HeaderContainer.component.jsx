import React from 'react';
import Header from './Header.component';
import MenuIcon from "./MenuButton.component";
import SearchBar from './SearchBar.component';
import ThemeSelector from './ThemeSelector.component';
import SessionButton from './SessionButton.component';

function HeaderContainer({onInputChange, inputText, disabled}){
    return(
        <Header>
            <MenuIcon></MenuIcon> 
            <SearchBar 
              initSearchQuery={inputText} 
              onChange={onInputChange}
              disabled={disabled}
            ></SearchBar>
            <ThemeSelector></ThemeSelector>
            <SessionButton></SessionButton>
        </Header>
    );
};

export default HeaderContainer;