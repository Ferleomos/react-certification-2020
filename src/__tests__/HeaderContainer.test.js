import React from 'react';
import { render, screen } from '@testing-library/react';
import 'jest-styled-components';
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom/extend-expect';
import Header from '../components/Header/Header.component';
import MenuIcon from '../components/Header/MenuButton.component';
import SessionButton from '../components/Header/SessionButton.component';
import ThemeSelector from '../components/Header/ThemeSelector.component';
import SearchBar from '../components/Header/SearchBar.component';
import HeaderContainer from '../components/Header/HeaderContainer.component';

describe('validate header container works properly', () => {
    test('render header contains elements', () => {
        const headerC = renderer.create(<HeaderContainer />).toJSON();
        expect(headerC.children.length).toBe(4);
    });
});

describe('validate header works properly', () => {
  test('render header is working', () => {
    const hdr = renderer.create(<Header />).toJSON();
        expect(hdr).toMatchSnapshot()
    });

    test('applies default styles', () => {
        const hdr = renderer.create(<Header />).toJSON();

        expect(hdr).toHaveStyleRule('min-height', '64px');
        expect(hdr).toHaveStyleRule('display', 'flex');
    });
});

describe('validate menu button renders', () => {

    test('render button properly', () => {
        render(<MenuIcon />);
        // screen.debug();
        const btnIcon = screen.getByRole('button');
        // screen.debug(btnIcon);
        expect(btnIcon).toBeInTheDocument();
    });

    test('render menu is working', () => {
        const menuBtn = renderer.create(<MenuIcon />).toJSON();
        expect(menuBtn).toMatchSnapshot();
    });
});

describe('validate session button works properly', () => {

    test('render session button is working', () => {
        const snButton = renderer.create(<SessionButton />).toJSON();
        expect(snButton).toMatchSnapshot();
    });

    test('applies default styles', () => {
        const snButton = renderer.create(<SessionButton />).toJSON();
        expect(snButton).toHaveStyleRule('display', 'inline-flex');
    });
});

describe('validate search bar renders', () => {

    test('render search bar works properly', () => {
        render(<SearchBar />);
        
        const inputSearch = screen.getByRole('textbox', { selector: 'input'});
        const placeholderText = screen.getByPlaceholderText(/Search/i);

        expect(inputSearch).toBeInTheDocument();
        expect(placeholderText).toBeInTheDocument();
    });
});

describe('validate theme selector works properly', () => {
    test('render theme selector button is working', () => {
        const selTheme = renderer.create(<ThemeSelector />).toJSON();
        expect(selTheme).toMatchSnapshot();
    });

    test('applies default styles', () => {
        const selTheme = renderer.create(<ThemeSelector />).toJSON();
        expect(selTheme).toHaveStyleRule('display', 'inline-flex');
    });
});


// /// ASYNC EXAMPLES /////

// const asyncCallback = (cb) => {
//     setTimeout(() => {
//         cb(true);
//     }, 1000);
// };

// const asyncPromise = () => new Promise((resolve) => resolve(true));

// test('async with callback', (done) => {
//     asyncCallback(() => {
//         expect(result).toBe(true);
//         done();
//     });
// });

// test('async with promises', () => {
//     return asyncPromise().then((result) => expect(result).toBe(true));
// });

// test('async with async await', async () => {
//     const result = await asyncPromise();
//     expect(result).toBe(true);
// });
