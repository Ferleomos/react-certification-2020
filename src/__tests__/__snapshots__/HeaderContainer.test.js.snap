// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validate header works properly render header is working 1`] = `
.c0 {
  background-color: #60a4e4;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  border: none;
  z-index: 1100;
  width: 100%;
  min-height: 64px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<header
  className="c0"
/>
`;

exports[`validate menu button renders render menu is working 1`] = `
.c1 {
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  width: 25px;
  height: 25px;
}

.c0 {
  background-color: transparent;
  color: transparent;
  border: none;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  position: relative;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin: 0px 25px 0px 15px;
}

@media (max-width:640px) {
  .c0 {
    margin: 0px 0px 0px 10px;
  }
}

<button
  className="c0"
>
  <svg
    className="c1"
    viewBox="0 0 24 24"
  >
    <path
      d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
    />
  </svg>
</button>
`;

exports[`validate session button works properly render session button is working 1`] = `
.c1 {
  fill: white;
}

.c0 {
  border: none;
  border-radius: 50%;
  height: 40px;
  min-width: 40px;
  background-color: #bdbdbd;
  color: transparent;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  margin: 0px 20px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  vertical-align: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-align-content: center;
  -ms-flex-line-pack: center;
  align-content: center;
}

@media (max-width:640px) {
  .c0 {
    display: none;
  }
}

<button
  className="c0"
>
  <svg
    className="c1"
    viewBox="0 0 24 24"
  >
    <path
      d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
    />
  </svg>
</button>
`;

exports[`validate theme selector works properly render theme selector button is working 1`] = `
.c0 {
  border: none;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  vertical-align: center;
  width: 100%;
  margin: auto;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 15px;
}

.c2 {
  color: Black;
  font-weight: bold;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
}

.c1 {
  margin: 0px 15px;
}

@media (max-width:640px) {
  .c0 {
    display: none;
  }
}

<div
  className="c0"
>
  <input
    className="c1"
    type="checkbox"
  />
  <label
    className="c2"
  >
    Dark Mode
  </label>
</div>
`;
